File: backend/models/instance.go
package models

import (
	"time"
	"github.com/globalsign/mgo/bson"
)

type Instances []*Instance

// an Instance is a slave to the coordinator which handles spammers on its host system.
//	- controls spammers local to it
//	- communicates with the coordinator for configuration
//	- supplies its health to the coordinator
type Instance struct {
	ID        bson.ObjectId `json:"id" bson:"_id"`
	Address   string        `json:"address" bson:"address"`
	APIToken  string        `json:"api_token" bson:"api_token"`
	Name      string        `json:"name" bson:"name"`
	Desc      string        `json:"desc" bson:"desc"`
	Tags      []string      `json:"tags" bson:"tags"`
	Online    bool          `json:"online" bson:"online"`
	CreatedOn time.Time     `json:"created_on" bson:"created_on"`
	UpdatedOn time.Time     `json:"updated_on" bson:"updated_on"`
}

File: frontend/js/models/Instance.ts
import {Collection, Model} from "./BackOn";

export class Instance extends Model {
    address: string;
    api_token: string;
    name: string;
    desc: string;
    tags: Array<string>;
    online: boolean;
    created_on: Date;
    updated_on: Date;

    constructor(attrs?: any) {
        super(attrs);
        this.url = '/api/instances/id';
    }
}

export class Instances extends Collection<Instance> {
    constructor(models?: Array<Instance>, options?: any) {
        super(Instance, models, options);
    }
}

export class AllInstances extends Instances {
    constructor(models?: Array<Instance>, options?: any) {
        super(models, options);
        this.url = `/api/instances`;
    }
}
